### PR TITLE
Residue rotation and UA translation axes for terminal residues

### PR DESCRIPTION
**Pull request by ioanaapapa**
## Summary
Terminal residue axes are no longer dependent on atom indices. The edge atom (i.e. heavy atom bonded to neighbour residue) is chosen alongside the average position of heavy atoms bonded to the edge atom and average position of all other heavy in the residue as the three points used for determining the axes. The centre of rotation (at res level)/translation (at UA level) and axes are then computed from the three points similar to the case of non-terminal residues. 

### Terminal residue treatment was changed
- Residue rotation and UA translation axes for terminal residues no longer use atom indices to emulate the backbone for non-terminal residues.
- The edge atom (E1 - red), defined as the heavy atom bonded to a neighbour residue is identified. All heavy atoms bonded to the edge atom (purple) are identified and their average position (C) is used as the third point for obtaining the centre of rotation. All other heavy atoms in the residue (green) are identified and used as the equivalent to the second edge(E2).
 
image

- The centre of rotation (O - blue) is the point where the perpendicular from the C (purple)  meets the E1 (red)-E2 (green) vector.
- x-axis is along the centre of rotation (O) - first edge(E1) vector
- y axis is along the centre of rotation (O) - average position of bonded atoms (C) vector
- z-axis is perpendicular to the two 

### Edge cases have been introduced
- If there are no other heavy atoms in a terminal residue, other than the edge atom and those bonded to the edge atom, the centre of rotation is the edge atom and axes are defined as: x-axis is along the edge-bonded vector, y axis is an arbitrary perpendicular axis and z-axis is perpendicular to the two. (This uses the get_custom_axes function and is the same as case2 in get_bonded_axes for UAs)
- If the two edge atoms in a non-terminal residue are bonded (there is no backbone atom between them), the centre of rotation is located at their average position and axes are defined as: x-axis is along the centre-of-rotation-first edge vector, y-axis is arbitrary perpendicular and z-axis is perpendicular to the two.

### Tests for terminal residues and edge cases 
- Unit tests for terminal residues have been updated to reflect changes.
- New tests for the edge cases covered have been introduced

### Documentation
- Vibrational entropy in theory section of the documentation was modified to reflect changes.  

### get_residue_axes and get_UA_axes structure change
- New functions get_terminal_axes and get_non_terminal_axes are introduced and called in get_residue_axes and get_UA_axes to improve code flow.

## Impact
- This closes issue #399.
- Terminal residue treatment is no longer dependent on atom indexing, which is arbitrary. The same result would be obtained using topologies of the same molecule with different indexes.
- Residue rovibrational results are now slightly higher compared to total entropy results obtained using the index-based backbone. No signficant difference observed for UA transvibrational results. (only treatment of terminal residues differs in table below)
image

### CODE REVIEW

File: CodeEntropy/levels/axes.py [hunk 1]
```diff
@@ -73,17 +73,35 @@ def get_residue_axes(
...
         - If there are *no* bonds to other residues:
             * Use a custom principal axes, from a moment-of-inertia (MOI) tensor
-              that uses positions of heavy atoms only, but including masses of
+              that uses positions of heavy atoms only, but includes masses of
               heavy atom + bonded hydrogens.
             * Set translational axes equal to rotational axes (as per the original
               code convention).
-        - If bonded to other residues:
+
+        - If bonded to only one other residue:
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py:81]: I wonder if it might be worth splitting the different cases into their own helper functions and calling them from the `if` statements? This function is getting quite large now, and there’s a fair bit of conditional logic to follow. I think separating the cases out could make the overall flow a bit clearer and easier to maintain as the logic grows.

File: CodeEntropy/levels/axes.py [hunk 2]
```diff
@@ -145,30 +163,51 @@ def get_residue_axes(
...
+                # find the average position of all other heavy atoms in residue
+                other_atoms = []
+                for atom in uas:
+                    if atom != edge_atom and atom not in bonded_atoms:
+                        other_atoms.append(atom)
+                if len(other_atoms) > 0:
+                    average_other_atoms = np.zeros(3)
+                    for atom in other_atoms:
+                        average_other_atoms += atom.position
+                    average_other_atoms /= len(average_other_atoms)
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: I think there may be a small issue here: `average_other_atoms` is a 3-element coordinate vector, so `len(average_other_atoms)` will always be 3 rather than the number of atoms being averaged. Should this be `len(other_atoms)`? This would affect the position used to determine the rotational axes.

File: CodeEntropy/levels/axes.py [hunk 3]
```diff
@@ -145,30 +163,51 @@ def get_residue_axes(
...
+                    rot_center, rot_axes = self.get_residue_custom_axes(
+                        [edge_atom.position, average_other_atoms], average_bonded_atom
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: Could we handle the case where `bonded_atoms` is empty here? `average_bonded_atom` is only initialised inside the `if len(bonded_atoms) > 0` block, but is used here.

**Review by harryswift01 — CHANGES_REQUESTED** [review]
Thank you for sending these changes through! I think this is a really good improvement, and it’s great to see the terminal residue handling becoming more robust and independent of atom indexing. I’ve just left a few comments around some small technical and maintainability points. If you’re able to have a look through them and let me know what you think, that would be great. Overall, I think this is looking really good and is very close to being ready to go in!

File: CodeEntropy/levels/axes.py (hunk 1 above)
Comment (ioanaapapa) [reply CodeEntropy/levels/axes.py:81]: This is a good idea as it would make the code easier to follow.

File: CodeEntropy/levels/axes.py (hunk 2 above)
Comment (ioanaapapa) [reply CodeEntropy/levels/axes.py]: Yes, that should be len(other_atoms). I'll check my results in case I used this and got the wrong positions.

**Comment by ioanaapapa** [comment]
@harryswift01 Thank you, Harry! I will make the changes.

**Comment by ioanaapapa** [comment]
@harryswift01 I made the changes. Hope it's okay now. I've also re-run the average position calculations with the correct average position and the result was the same as the table above.

File: CodeEntropy/levels/axes.py [hunk 4]
```diff
@@ -598,7 +559,99 @@ def get_residue_custom_axes(self, edges, center):
...
+                average_other_atoms = np.zeros(3)
+                for atom in other_atoms:
+                    average_other_atoms += atom.position
+                average_other_atoms /= len(other_atoms)
+                rot_center, rot_axes = self.get_residue_custom_axes(
+                    [edge.position, average_other_atoms], average_bonded
+                )
+            else:
+                rot_center = edge.position
+                rot_axes = self.get_custom_axes(
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py:649]: Just something to double check, this calls `get_custom_axes(a=edge.position, b=[average_bonded], c=np.zeros(3))`, but the real signature is `(a, b_list, c, dimensions)`, so `b` should be `b_list`, and dimensions is missing entirely. This will throw a `TypeError` at runtime for any terminal residue with exactly 2 heavy atoms. Probably easiest to add dimensions as a parameter on `get_terminal_axes` and pass `data_container.dimensions[:3]` through from the callers.

File: CodeEntropy/levels/axes.py [hunk 5]
```diff
@@ -598,7 +559,99 @@ def get_residue_custom_axes(self, edges, center):
...
+        if len(backbone) > 0:
+            for heavy_atom in backbone:
+                backbone_center += heavy_atom.position
+            backbone_center /= len(backbone)
+            rot_center, rot_axes = self.get_residue_custom_axes(
+                edges.positions, backbone_center
+            )
+        else:
+            rot_center = (edges[0].position + edges[1].position) / 2
+            rot_axes = self.get_custom_axes(a=rot_center, b=[edges[0]], c=np.zeros(3))
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: Same kind of issue as the one above, this calls `get_custom_axes(a=rot_center, b=[edges[0]], c=np.zeros(3))`, but again `b` should be `b_list`, and dimensions is missing. It also passes `edges[0]` directly, an atom, rather than `edges[0].position`, which `get_custom_axes` needs.

File: tests/unit/CodeEntropy/levels/test_axes.py [hunk 6]
```diff
@@ -1684,20 +1624,419 @@ def test_get_UA_axes_raises_when_only_rot_axes_fail(monkeypatch):
...
+    )
+
+    def _select_atoms(q):
+        if q == "mass 2 to 999":
+            return heavy_atoms
+        if q.startswith("(mass 2 to 999) and bonded"):
+            return [heavy_atoms[1]]
+
+    residue.atoms.select_atoms.side_effect = _select_atoms
+    monkeypatch.setattr(ax, "get_custom_axes", lambda a, b, c: 2 * np.eye(3))
```
Comment (harryswift01) [inline tests/unit/CodeEntropy/levels/test_axes.py]: I think this test is passing because it's mocking the wrong part of `get_custom_axes` (not quite matching the signature we talked about above). If we tidy up the mock to match, I reckon it'd catch the bug above nicely.

**Review by harryswift01 — CHANGES_REQUESTED** [review]
This looks like its shaping up really nicely, I think the new smaller functions make this a lot nicer and easier to follow so thank you for doing this. I have added a few more comments just to double check but nothing major. One of my comments regarding the testing I noticed in a few of the tests (`test_get_non_terminal_axes_2points`, `test_get_ua_axes_bonded_terminal_2_points`, `test_get_ua_axes_non_terminal_2_atoms`, `test_get_residue_axes_terminal_2_atoms`, `test_get_residue_axes_non_terminal_2_atoms`) I think it would be great if they could all follow the same approach, let me know if there is any way I can help but this is looking really good!

**Comment by ioanaapapa** [comment]
Thanks, @harryswift01! Is your last comment about the test mocking the wrong part of get_custom_axes what you were referring to as an issue in multiple tests or is there a second issue, as well?

**Comment by harryswift01** [comment]
> Thanks, @harryswift01! Is your last comment about the test mocking the wrong part of get_custom_axes what you were referring to as an issue in multiple tests or is there a second issue, as well?

Hi @ioanaapapa, thanks for double checking! Yes, this would be the same issue, not a second one, I just spotted it in a few other tests too. If you're able to apply the same fix there as well, that'd be great.

**Comment by ioanaapapa** [comment]
Thank you! I'll change the tests

File: CodeEntropy/levels/axes.py [hunk 7]
```diff
@@ -73,17 +73,36 @@ def get_residue_axes(
           (previous/next in sequence) using MDAnalysis bonded selections.
         - If there are *no* bonds to other residues:
             * Use a custom principal axes, from a moment-of-inertia (MOI) tensor
-              that uses positions of heavy atoms only, but including masses of
+              that uses positions of heavy atoms only, but includes masses of
+              that uses positions of heavy atoms only, but includes masses of
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py:76]: Looks like this might be a repeat on line 76 and 77

File: CodeEntropy/levels/axes.py [hunk 8]
```diff
@@ -73,17 +73,36 @@ def get_residue_axes(
...
               heavy atom + bonded hydrogens.
             * Set translational axes equal to rotational axes (as per the original
               code convention).
-        - If bonded to other residues:
+
+        - If bonded to only one other residue:
+
+        - If bonded to only one other residue:
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py:81]: Looks like this might also be a repeat on line 82 and 84

File: CodeEntropy/levels/axes.py [hunk 9]
```diff
@@ -73,17 +73,36 @@ def get_residue_axes(
...
+            * Translational axes are principal axes of data_container.
+            * Find edge heavy atom (i.e. heavy atoms bonded to neighbour residue).
+              Compute rotation centre and axes as in get_terminal_axes.
+              custom MOI, using heavy atom positions and heavy atom + hydrogen masses.
+
+        - If bonded to at least two other residues:
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: Looks like this block has been repeated within this docstring

File: CodeEntropy/levels/axes.py [hunk 10]
```diff
@@ -290,71 +300,52 @@ def get_UA_axes(self, data_container, index: int, res_position):
...
-                backbone_center = backbone_center / len(backbone)
-
-                trans_center, trans_axes = self.get_residue_custom_axes(
-                    edges, backbone_center
-                )
-                residue_heavy_atoms = residue.atoms.select_atoms("mass 2 to 999")
-
             # look for heavy atoms in residue of interest
+            residue_heavy_atoms = residue.atoms.select_atoms("mass 2 to 999")
+            residue_heavy_atoms = residue.atoms.select_atoms("mass 2 to 999")
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: These two lines also repeat here as well, lines 347 and 348

File: CodeEntropy/levels/axes.py [hunk 11]
```diff
@@ -598,7 +589,109 @@ def get_residue_custom_axes(self, edges, center):
         y_axis /= np.linalg.norm(y_axis)
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: Whilst I was having a look, I saw that `get_custom_axes` guards against zero-length axis vectors and raises a clear `ValueError` if they end up with zero length, but `get_residue_custom_axes` doesn't have the same check. I think it would be nice to put that raise check in here too, in case any NaN values creep in.

File: CodeEntropy/levels/axes.py [hunk 12]
```diff
@@ -247,18 +252,23 @@ def get_UA_axes(self, data_container, index: int, res_position):
...
+            - If there are *no* bonds to other residues, use a custom principal axes
+            from a moment-of-inertia (MOI) tensor that uses positions of heavy atoms
+            only, but includes masses of heavy atom + bonded hydrogens.
+            - If bonded to only one other residue, see get_terminal_axes.
+            - If bonded to at least two other residues, see get_non_terminal_axes.
+            - If there are *no* bonds to other residues, use a custom principal axes
+            from a moment-of-inertia (MOI) tensor that uses positions of heavy atoms
+            only, but includes masses of heavy atom + bonded hydrogens.
+            - If bonded to only one other residue, see get_terminal_axes.
+            - If bonded to at least two other residues, see get_non_terminal_axes.
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: Is this part duplicated or is this a difference case?

File: CodeEntropy/levels/axes.py [hunk 13]
```diff
@@ -247,18 +252,23 @@ def get_UA_axes(self, data_container, index: int, res_position):
...
 
         - Rotational axes:
             Identify heavy atoms in the residue/molecule of interest and choose
             the `index`-th heavy atom (where index corresponds to the bead index).
             Use bonded topology around that heavy atom to determine UA rotational
-            axes (see :meth:`get_bonded_axes`).
-            Compute a custom MOI tensor using heavy-atom coordinates but UA masses
-            (heavy + bonded H masses), then compute the principal axes from it.
+            axes (see :meth:`get_bonded_axes`). Compute a custom MOI tensor.
+            axes (see :meth:`get_bonded_axes`). Compute a custom MOI tensor.
```
Comment (harryswift01) [inline CodeEntropy/levels/axes.py]: Looks like this part is also a repeat of the previous line, for lines 270 and 271

**Review by harryswift01 — CHANGES_REQUESTED** [review]
Thanks again for making these changes @ioanaapapa it's looking really good now! Just a couple of small things I noticed, mostly around the docstrings: there are a few repeated lines in there that would be great if we tidied up, I've left comments pointing to each one. I also spotted one thing in `get_residue_custom_axes` it might be worth adding a guard there to catch zero-length axis vectors, similar to what `get_custom_axes` already does, just so we don't end up with any NaN values slipping through.

Other than that, this is very close, just a few minor polishing points and I think we'll be good to merge.

File: CodeEntropy/levels/axes.py (hunk 7 above)
Comment (ioanaapapa) [reply CodeEntropy/levels/axes.py:76]: Oh I think that might have happened when rebasing. Thank you so much for noticing!

**Comment by ioanaapapa** [comment]
@harryswift01 I've tidied up the docstrings and added a ValueError raise in get_residue_custom_axes, alongside a unit test for that. Hope it's okay now!

**Review by harryswift01 — APPROVED** [review]
Hi @ioanaapapa, thank you for all your work on this! This is all looking good from a technical side now and I can't see any reason why this can't be merged in. Just before we do merge this in, as this is a change in the numerics/science, if @skfegan or @jkalayan or @jimboid  could have also have a look at this as well just to make sure there isn't anything I've missed from the science side, then after their approval I'd be happy to see this merged in, this is all great work, thank you again!

**Review by jimboid — APPROVED** [review]
Improvements to the way terminal residues are treated will help to fix some edge case issues reported on the issue tracker. This has been discussed at length at the all hands meetings with Richard, happy to see this merged @ioanaapapa.
